Assets/Minigames/Fight/Scripts/Managers/RoomManager.cs

```diff
@@ -1,7 +1,173 @@
+        {
+            _roomSettings = GameManager.SettingsManager.progressSettings.CurrentWorld.RoomSettings;
+
+            List<RoomController> availableRooms = new();
+
+            // Instantiate start room.
+            var startRoom = Instantiate(_roomSettings.startRoom);
+
+            // Add start room to our list of rooms to branch from.
+            availableRooms.Add(startRoom);
+            var targetRoom = startRoom;
+
+            // Deterimine how many rooms we'll spawn.
+            int roomCount = Random.Range(_roomSettings.minRooms, _roomSettings.maxRooms + 1);
+
+
+            List<RoomController> roomsToInstantiate = new();
+            Vector2 direction = Vector2.zero;
+
+            // Populate our rooms to instantiate list with the determined number of random rooms
+            for (int i = 0; i < roomCount; i++)
+            {
+                roomsToInstantiate.Add(_roomSettings.GetRandomRoom());
+            }
+
+            // Generate the map
+            foreach (RoomController roomToInstantiate in roomsToInstantiate)
+            {
+                bool valid = false;
+                var room = roomToInstantiate;
+
+                Vector2 center = Vector2.zero;
+
+                // Repeat until valid placement has been found
+                while (!valid)
+                {
+                    // Select a random room to branch off of.
+                    int roomToBranchOffOf = Random.Range(0, availableRooms.Count);
+                    targetRoom = availableRooms[roomToBranchOffOf];
+
+                    // Select random direction to branch in.
+                    direction = GetRandomCardinalDirection();
+
+                    // Calculate the location the new room will spawn in.
+                    float x = (targetRoom.col.bounds.extents.x + (room.Tilemap.cellBounds.AsVector2().x / 2)) * direction.x;
+                    float y = (targetRoom.col.bounds.extents.y + (room.Tilemap.cellBounds.AsVector2().y / 2)) * direction.y;
+                    center = targetRoom.col.bounds.center.AsVector2() + new Vector2(x, y);
+
+
+                    // Abort before doing the overlap check if the chosen direction has already been used (slight performance boost).
+                    RoomConnection roomConnection = null;
+                    foreach (RoomConnection connection in targetRoom.roomConnections)
+                    {
+                        if (connection.Direction == direction)
+                        {
+                            if (connection.HasConnection)
+                            {
+                                continue;
+                            }
+                            roomConnection = connection;
+                        }
+                    }
+
+                    // Check perposed room location to ensure it won't overlap with an existing room
+                    if (!Physics2D.OverlapBox(center, room.Tilemap.cellBounds.AsVector2() * 0.99f, 0))
```

Comment:
we don't like magic numbers, make this a constant. maybe call it `collisionScaleFactor`